### ISSUE TITLE
Make OnUpload able to return a promise

### ISSUE THREAD

**Issue by mmezzawi** [FineUploader/fine-uploader#1760]
## Type of issue
- [ ] Bug report
- [X] Feature request
- [X] Support request

## Uploader type
- [X] Traditional
- [ ] S3
- [ ] Azure

#### Fine Uploader version
{5.13.0}

<details>
Dear @rnicholus, 

I have faced a case where i need to send a request to the sever once the upload start and wait the server response before the chunks or the file start upload, but seems that onUpload callback doesn't support promise.
So my questions are : 
1- is that currently available in another way ? 
2- is there any future plan to implement promise in onUpload ? 
3- Currently the now function is calling "options.onUpload(id, name);" 
so if we need to wait any promise to be resolved we should do it there. 
is there any consequence of doing that in now function?

Thanks, appreciate your support!!!
</details>

**Comment by rnicholus** [comment]
Can you provide more detail? Why do you need to send a request to the server before the upload starts? What is the nature of this request? Can it not be sent in your onSubmit handler?

**Comment by zaryk** [comment]
+1 - In my case, I would need to prompt the user if the file exists, and each file could have a prompt.  Also, with the work around that I provided below, users may complain about additional clicks.  Since the file already exists, there are 4 scenarios that could occur.  

- User can overwrite the file
- User can rename the file automatically, server would append _1 to the name
- User can resume the file, maybe because the cache has been removed already and the file on the server is smaller than the uploaded file (clearing history seems to remove it, even if it has not been expired).  Also, it is possible that for whatever reason, one person sees that the file has not been uploaded completely, it has expired, and wants to continue the upload.
- User can cancel and rename the file manually.  Event would have to be cancellable.

I have been able to work around this by checking if the file already exists onsubmit through ajax call (because I still want the file to show) and get the server filesize, and display a link for each file that already exists.  I modified the fineuploader with another status of AlreadyExists, and set the status to this so the other events do not trigger on it. 

NOTE: javascript is not my area of strength:

`
onSubmitted: function (id, name) {
                    //Triggered on drag and drop
                    var file = manualUploader.getFile(id);

                    var result = $.ajax({
                        type: "GET",
                        url: '/server/upload/exists?foldername=' + (file.qqPath === undefined ? "" : file.qqPath) + '&filename=' + file.name + "&size=" + manualUploader.getSize(id),
                        async: false,
                        cache: false
                    }).responseText;

                    var result = jQuery.parseJSON(result);
                    var file = $(".qq-file-id-" + id);
                    var selector = file.find("#duplicate-selector");
                    if (result.success === false) {
                        

                        file.find("#newSizeSelectorValue").val(result.newSize);
                        file.find("#oldSizeSelectorValue").val(result.oldSize);

                        var duplicate = file.find("#duplicate-options");
                        duplicate.click(function () {
                            $("input[name=fileoperations][value=4]").prop('checked', true);
                            $("#fileNamePlaceholder").text(name);

                            var file = $(".qq-file-id-" + id);
                            var oldSize = file.find("#oldSizeSelectorValue").val();
                            var newSize = file.find("#newSizeSelectorValue").val();
                            if (oldSize != "0B (0KB)" && oldSize !== newSize && Number(oldSize.substring(0, oldSize.indexOf("B"))) < Number(newSize.substring(0, newSize.indexOf("B")))) {
                                $("#resumePlaceholder").show();
                                $("#oldSizePlaceholder").text(oldSize);
                                $("#newSizePlaceholder").text(newSize);
                            }
                            else {
                                $("#resumePlaceholder").hide();
                            }


                            $("#duplicateId").val(id);
                            $("#duplicateModal").modal({                    // wire up the actual modal functionality and show the dialog
                                "backdrop": "static",
                                "keyboard": true,
                                "show": true                     // ensure the modal is shown immediately
                            });
                        });

                        
                        selector.show();

                        manualUploader.setStatus(id, qq.status.ALREADYEXISTS);
                    }
                    else {
                        selector.hide();
                    }
                },
`




`
onUpload: function (id, name) {
                    var file = $(".qq-file-id-" + id);

                    var selectorValue = file.find("#duplicateSelectorValue").val();
                    if (selectorValue !== "") {
                        manualUploader.setParams({ operation: selectorValue }, id);
                    }
                    else
                    {
                        manualUploader.setParams({ guid: "", operation: "" }, id);
                    }

                    if(selectorValue === "append")
                    {
                        var file = $(".qq-file-id-" + id);
                        var oldSize = file.find("#oldSizeSelectorValue").val();
                        var newSize = file.find("#newSizeSelectorValue").val();
                        manualUploader.updateProgress(id, name, oldSize.substring(0, oldSize.indexOf("B")), newSize.substring(0, newSize.indexOf("B")));
                    }
                },
`

**Comment by rnicholus** [comment]
>I modified the fineuploader with another status of AlreadyExists, and set the status to this so the other events do not trigger on it.

Why not just return a `Promise` in your `onSubmit` and reject it if the file already exists?

**Comment by zaryk** [comment]
I want the file to still show in the uploader.  If I am not mistaken, rejecting on the onSubmit removes the file from FineUpload files so it no longer shows.  If it no longer shows, then there is no longer a prompt for the user to be able to handle it within fineuploader.  If there is no prompt, then there is no business login that can be shown to the user.

**Comment by mmezzawi** [comment]
Dear @rnicholus , 

In my case , i need to send a request to a storage server that will execute the following : 
1- create path directory 
2- grant user access to the directory
3- send a request to server 'B' using a paid service to log the upload and get a signature 
( in case 'B' return a false we need to run another part of business )
4- send the signature back to the client side 
5- upload will start if all is ok

I'm not using onSubmit because the process should start when user start the upload not just submit the file to the upload queue.


Thanks.

**Comment by mmezzawi** [comment]
Dear @rnicholus , 

Another question , is there a way to get the chunk info data in onupload  ? 
if not, how to check if file will be upload in chunk, since as i read in the documentation that if file size is less then the chunk size it will not be sent in chunk mode ( which is logic ) .
In our case we need to find out if the file is sent as chunk or as a complete file .

Thanks .

**Comment by rnicholus** [comment]
You can determine the file size using the [`getSize()` API method](https://docs.fineuploader.com/branch/master/api/methods.html#getSize). From there, you can determine if the file is large enough to chunk.

Let's please keep this thread/issue on topic. If there are other questions/concerns, please open up a new issue.

**Comment by mmezzawi** [comment]
Dear @rnicholus ,

concurrent upload is another reason to have onUpload return promise , since it will be responsible of creating all the need folders and once everything is done upload can start , else the user will need to create lock on server to make sure that only the first arriving request will create the folder.
Using the existence of the folder as a flag to not create one will not be a good idea, since two request could be server at the same time.

Thanks.

**Comment by rnicholus** [comment]
>the user will need to create lock on server to make sure that only the first arriving request will create the folder

Your server should be able to account for this by tracking appropriate file/folder state in a database, for example.

**Comment by rnicholus** [comment]
duplicate of #848
